### DISCUSSION TITLE
C++ client - multipart upload example.

### DISCUSSION THREAD

**Discussion by epasveer** [googleapis/google-cloud-cpp#14295] [Q&A] [answered]
Does anyone have an example to upload a local file to a bucket using "multipart" or "parallel composite"?

Basically trying to upload a large local file. Having the C++ sdk automatically breakup the file into parts makes sense.
I'm not finding any good examples out there on the web.

Thanks.

**Comment by alevenberg** [comment]
To clarify, are you looking for an example like [this](https://cloud.google.com/storage/docs/parallel-composite-uploads#storage-upload-object-pcu-java) in C++?

I can try and create one.

**Reply by alevenberg** [reply, marked as answer]
Actually, is this what you are looking for  https://github.com/GoogleCloudPlatform/cpp-samples/blob/main/gcs-fast-transfers/upload.cc?

**Reply by epasveer** [reply]
Yes, exactly.

With Amazon, you can create a multipart transfer pretty easy. You just tell the client to break them up in "parts", and you specify the part size.

I just tried chatgpt and it gave a result where I need to use:
```
  auto res = client.UploadChunk(
  auto res = client.ComposeObject(request);
  google::cloud::storage::ComposeObjectRequest
```
Which basically means I need to implement my own "multipart" solution using GCP low level api's.

If so, I guess that's okay. I just thought it would be built into the C++ sdk.  Maybe other language sdk's already have it.

Also, thanks for your quick response!

**Reply by alevenberg** [reply]
Hmm... I don't know a lot about the storage API, and our expert is OOO atm. I asked someone else to see what the status of the API is. I will update the question when I find an answer.

**Reply by epasveer** [reply]
About the code I got from chatgpt, it doesn't compile and I grep'ed in the c++ sdk for thos function names and they don't exist.

So chatgpt must have scrapped an old solution from some time ago...

**Reply by epasveer** [reply]
> Actually, is this what you are looking for https://github.com/GoogleCloudPlatform/cpp-samples/blob/main/gcs-fast-transfers/upload.cc?

This looks better!  I'll check it out.

**Reply by epasveer** [reply]
With the suggestions from @alevenberg , I've also found this other link.

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/storage/examples/storage_object_file_transfer_samples.cc

I'll report back tomorrow with my results...

**Reply by alevenberg** [reply]
Update: we do not support multipart-uploads (https://github.com/googleapis/google-cloud-cpp/issues/7054). I think the example I added should cover your use case though.

**Reply by epasveer** [reply]
Thanks again.  I'll try your example and will let you know.

**Reply by epasveer** [reply]
Your example works well.  Thank you!

This issue can be closed.

**Reply by alevenberg** [reply]
Awesome! Glad it worked for you!
